docs: release notes for Renovate v39 (#32504)
renovatebot/renovate: docs/usage/release-notes-for-major-versions.md

This is because recent versions may revert changes made in an older version.
You also don't have to scroll to the bottom of the page to find the latest release notes.

## Version 39

### Breaking changes for 39

#### New tools and config for Renovate's Docker images

Renovate Docker images now use Node.js v22 as base, this was Node.js v20.

##### Sidecar images

Renovate's "sidecar" images now:

- Use Ubuntu 24.04 as base, this was 20.04

> - Set the user ID to `12021`, this was `1001`

Reviewer: All renovate images are now having this, they are all extend the sidecar image. so move one heading up

Reviewer: Also add a note that caches need to be removed or perms need to be fixed because of the userid change.

##### Full images

Renovate's `-full` images now uses:

- Ubuntu 24.04
- Python 3.13
- Node.js v22

The Renovate `-full` image defaults to [`binarySource=global`](https://docs.renovatebot.com/self-hosted-configuration/#binarysource).
If you want to keep the old behavior, where Renovate dynamically installs the needed tools: set the environment variable `RENOVATE_BINARY_SOURCE` to `"install"`.

Question do we need to update the `binarySource` docs?
It still says:

> > Starting in v36, Renovate's default Docker image (previously referred to as the "slim" image) uses `binarySource=install` while the "full" Docker image uses `binarySource=global`. If you are running Renovate in an environment where runtime download and install of tools is not possible then you should use the "full" image.

Reviewer: this was a bug in v38 where we moved the docker build back to this repo. the following line got's lost:

https://github.com/renovatebot/docker-renovate-full/blob/6c5fd24430118211b68db8eac7e9395a7e8716bb/Dockerfile#L17

#### Prefer squash merges for automerges on GitHub

When automerging, Renovate tries the "squash merge" method first.

#### Branch names with multiple slashes

Branch names with multiple forward slashes (`/`) will change if `branchNameStrict=true`.

### Commentary for 39

#### Prefer squash merge for automerges to allow signed commits on GitHub

This info is for the GitHub platform only.

When you sign your commits on GitHub, you'll want Renovate's commits to be signed too.
If Renovate automerges its PRs with the **Rebase and Merge** option, then GitHub can't sign the commits.
But GitHub can sign squash-merged commits for Renovate.

> This is why we changed the default behavior to try the squash merge first.

Author: Please check if this is technically correct!

Read the [GitHub Docs, Signature verification for rebase and merge](https://docs.github.com/en/authentication/managing-commit-signature-verification/about-commit-signature-verification#signature-verification-for-rebase-and-merge) to learn more.

Note from HonkingGoose to maintainers, the [GitHub Docs, signature verification for bots](https://docs.github.com/en/authentication/managing-commit-signature-verification/about-commit-signature-verification#signature-verification-for-bots) have more info about signing commits as bot author.

> I don't know if we already do this, or if the method is better than relying on a specific kind of merge type?

Author: Are we already doing this? I wonder if this may help avoid things like relying on a specific merge method to keep Renovate's commits signed?

> ### Signature verification for bots
>
> Organizations and GitHub Apps that require commit signing can use bots to sign commits. If a commit or tag has a bot signature that is cryptographically verifiable, GitHub marks the commit or tag as verified.
>
> Signature verification for bots will only work if the request is verified and authenticated as the GitHub App or bot and contains no custom author information, custom committer information, and no custom signature information, such as Commits API.

#### Why we we change branch names with multiple slashes

> Explain why this was changed here.

Author: Needs to be expanded by a maintainer.

This is because recent versions may revert changes made in an older version.
You also don't have to scroll to the bottom of the page to find the latest release notes.

## Version 39

### Breaking changes for 39

#### New tools for all Docker images

All our Docker images now use:

- Node.js v22 as base, was Node.js v20
- Ubuntu 24.04 as base, was 20.04

#### New Docker user ID for all Docker images

All our Docker images now set the Docker user ID to `12021`, the old ID was `1001`.

After updating your Renovate Docker image to the new v39 release, you must:

- Delete your old Docker cache, _or_
- Ensure the new user ID has write permissions to any existing cache

#### Updated version of Python, and new default behavior for the `-full` Docker image

On top of the changes listed above, the `-full` image now:

- Uses Python 3.13
- Defaults to [`binarySource=global`](https://docs.renovatebot.com/self-hosted-configuration/#binarysource) (note: this was previously the case in v36 onwards but regressed sometime in v38).

If you want to keep the old behavior, where Renovate dynamically installs the needed tools: set the environment variable `RENOVATE_BINARY_SOURCE` to `"install"`.

> 

Reviewer 2: ```suggestion
```

#### Renovate tries squash merges first when automerging on GitHub

Due to technical reasons, GitHub will only sign commits coming from a squash merge.
To help those who want Renovate to sign its commits, Renovate now tries the squash merge first.

Of course, Renovate only uses the merge method(s) that you allow in your GitHub repository config.

##### How you can allow squash merges on your GitHub repository

If you want to allow squash merges on your GitHub repository, follow the steps in the [GitHub Docs, configuring commit squashing for pull requests](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/configuring-commit-squashing-for-pull-requests).

#### Branch names with multiple slashes

> Branch names with multiple forward slashes (`/`) will change if `branchNameStrict=true`.

Reviewer 2: ```suggestion
Branch names with multiple forward slashes (`/`) will change if `branchNameStrict=true`.
Previously there were cases where special characters could still enter branch names despite `branchNameStrict=true` being set.
This has now been corrected, which can result in some branch name changes.
```

This is because recent versions may revert changes made in an older version.
You also don't have to scroll to the bottom of the page to find the latest release notes.

## Version 39

### Breaking changes for 39

#### New tools for all Docker images

All our Docker images now use:

- Node.js v22 as base, was Node.js v20
- Ubuntu 24.04 as base, was 20.04

#### New Docker user ID for all Docker images

All our Docker images now set the Docker user ID to `12021`, the old ID was `1001`.

After updating your Renovate Docker image to the new v39 release, you must:

- Delete your old Docker cache, _or_

> - Cut and paste the permissions from the old ID into the new user ID

Reviewer: ensure the new user id has write permission to the cache

> - Cut and paste the permissions from the old ID into the new user ID

Reviewer 2: ```suggestion
- Ensure the new user ID has write permissions to any existing cache
```

#### Updated version of Python, and new default behavior for the `-full` Docker image

On top of the changes listed above, the `-full` image now:

- Uses Python 3.13

> - Defaults to [`binarySource=global`](https://docs.renovatebot.com/self-hosted-configuration/#binarysource).

Reviewer 2: ```suggestion
- Defaults to [`binarySource=global`](https://docs.renovatebot.com/self-hosted-configuration/#binarysource) (note: this was previously the case in v36 onwards but regressed sometime in v38).
```

If you want to keep the old behavior, where Renovate dynamically installs the needed tools: set the environment variable `RENOVATE_BINARY_SOURCE` to `"install"`.

> Question do we need to update the `binarySource` docs?
> It still says:
> 
> > Starting in v36, Renovate's default Docker image (previously referred to as the "slim" image) uses `binarySource=install` while the "full" Docker image uses `binarySource=global`. If you are running Renovate in an environment where runtime download and install of tools is not possible then you should use the "full" image.

Reviewer 2: ```suggestion
```
